File: src/main/java/com/miueon/leetcode/editor/en/1011capacityToShipPackagesWithinDDays.ts
/*
 * @lc app=leetcode id=1011 lang=typescript
 *
 * [1011] Capacity To Ship Packages Within D Days
 */

// @lc code=start
function shipWithinDays(weights: number[], D: number): number {
    let left = getMax(weights)
    let right = getSum(weights)
    let mid = 0
    while(left <= right){
        mid = left + (right - left) / 2
        if(canFinish(weights, D, mid)){
            right = mid -1
        } else {
            left = mid + 1
        }
    }
    return left
};
function getMax(nums:number[]):number{
    let max = nums[0]
    for (let i of nums) {
        if(i > max){
            max = i
        }
    }
    return max
}

function getSum(n:number[]):number{
    let count = 0
    n.forEach((value) => {
        count += value
    })
    return count
}
function canFinish(weights:number[], D: number, cap:number):Boolean{
    let i = 0
    for (let day = 0; day < weights.length; day++) {
        let maxCap = cap
        while((maxCap -= weights[i])>= 0) {
            i++
            if(i == weights.length) {
                return true
            }
        }       
    }
    return false
}
// @lc code=end

File: src/main/java/com/miueon/leetcode/editor/en/1011CapacityToShipPackagesWithinDDays.kt
/*
 * @lc app=leetcode id=1011 lang=kotlin
 *
 * [1011] Capacity To Ship Packages Within D Days
 */
// leetcode submit region begin(Prohibit modification and deletion)
// @lc code=start
class Solution {
    fun shipWithinDays(weights: IntArray, D: Int): Int {
        var left = getMax(weights)
        var right = getSum(weights) + 1
        var mid = 0
       while (left <= right) {
           mid = left + (right - left) / 2
           if (canFinish(weights, D, mid)) {
              right = mid - 1
           } else {
               left = mid + 1
           }
       }
       return left

   }

   fun getMax(w: IntArray): Int {
       var max = w[0]
       for (i in w) {
           max = Math.max(i, max)
       }
       return max
   }

    fun getSum(w: IntArray): Int {
        var sum = 0
        w.forEach { sum += it }
        return sum
    }

    fun canFinish(w: IntArray, D: Int, cap: Int): Boolean {
        //        var i = 0
        //        for (day in 0 until D) {
        //            var maxCap = cap
        //            maxCap -= w[i]
        //            while (maxCap >= 0) {
        //                i++
        //                if (i == w.size) {
        //                    return true
        //                }
        //                maxCap -= w[i]
        //            }
        //        }
        //        return false
        var i = 0
        for (day in 0 until D) {
            var maxCap = cap
            while (w[i].let {
                maxCap -= it
                maxCap
            } >= 0) {
                i++
                if (i == w.size) return true
            }
        }
        return false
    }
}
// @lc code=end
// leetcode submit region end(Prohibit modification and deletion)
